JavaScript/chapter03/p01_three_in_one/avc278.js

```diff
@@ -0,0 +1,164 @@
+                return poppedVal;
+            case 3:
+                if (this.arr[this.stackThreeCurrIdx] === null) return;
+                poppedVal = this.arr[this.stackThreeCurrIdx];
+                this.arr[this.stackThreeCurrIdx] = null;
+                this.stackThreeCurrIdx = Math.max(this.stackThreeCurrIdx - 1, this.stackThreeStartIdx);
+                return poppedVal;
+            default:
+                return;
+        }
+    };
+    peek(stackNum) {
+        switch (stackNum) {
+            case 1:
+                return this.arr[this.stackOneCurrIdx];
+            case 2:
+                return this.arr[this.stackTwoCurrIdx];
+            case 3:
+                return this.arr[this.stackThreeCurrIdx];
+            default:
+                return;
+        }
+    };
+    push(stackNum, val) {
+        switch (stackNum) {
+            case 1:
+                if (this.stackOneCurrIdx === this.stackTwoStartIdx - 1) return;
+                if (this.arr[this.stackOneCurrIdx] !== null) this.stackOneCurrIdx += 1;
+                this.arr[this.stackOneCurrIdx] = val;
+                break;
+            case 2:
+                if (this.stackTwoCurrIdx === this.stackThreeStartIdx) return;
+                if (this.arr[this.stackTwoCurrIdx] !== null) this.stackTwoCurrIdx += 1;
+                this.arr[this.stackTwoCurrIdx] = val;
+                break;
+            case 3:
+                if (this.stackThreeCurrIdx === this.arr.length) return;
+                if (this.arr[this.stackThreeCurrIdx] !== null) this.stackThreeCurrIdx += 1;
+                this.arr[this.stackThreeCurrIdx] = val;
+                break;
+            default:
+                return;
+        }
+    };
+};
+
+// Helper function to convert a stack from A -> B -> C to C -> B -> A, and return its length
+const invertStackAndGetLength = (stack) => {
+    const res = new Stack();
+    let length = 0;
+    while (!stack.isEmpty()) {
+        length += 1;
+        res.push(stack.pop())
+    }
+    return [ res, length ];
+}
+
+/**
+ * @param   {Stack} A input stack
+ * @param   {Stack} B input stack
+ * @param   {Stack} C input stack
+ * @return  {_Array}  the array containing three stacks, with stack methods
+ *
+ * In implementing threeInOne, we need to know the longest stack length, which we'll call N, and the others M and P.
+ * As we traverse through these stacks, we do so one after the other, which results in O(N + M + P), which is really
+ * just O(N) as we can remove smaller terms. Creating the inverse stacks requires a maximum of O(N) space for the same
+ * reason.
+ * Now, when creating the output array (of type _Array), we require 3 * N additional time and space for traversal.
+ * Again, however, this results in O(N) time and O(N) space as we can ignore the leading term.
+ * 
+ * Runtime: O(N)
+ * Space:   O(N)
+ *
+ */
+const threeInOne = (A, B, C) => {
+    const [ invA, lenA ] = invertStackAndGetLength(A);
```

Comment:
normally, queue and stack implementations have an O(1) `.length()`/`.size()` operation (e.g. https://en.cppreference.com/w/cpp/container/stack/size). and an array-to-stack-then-inverted is essentially a queue, so some of the complexity needed to set this up can be avoided.

Reply:
Oh interesting!
I'll have to add the `size()` method to the library file in a future commit.
I'm wondering how I would solve this problem without the use of the queue class. Only using the given stacks, and the arrays.... Maybe as I'm pushing elements from the stacks into the arrays, I do so from the end to the start! That might work and resolve the issue of the unnecessary added complexity. I'll try that and send it in a future commit